fdbcli/CoordinatorsCommand.actor.cpp

```diff
@@ -0,0 +1,184 @@
+	StringRef nameTokenBegin = LiteralStringRef("description=");
+	for (auto tok = tokens.begin() + 1; tok != tokens.end(); ++tok)
+		if (tok->startsWith(nameTokenBegin)) {
+			new_cluster_description = tok->substr(nameTokenBegin.size());
+			std::copy(tok + 1, tokens.end(), tok);
+			tokens.resize(tokens.size() - 1);
+			break;
+		}
+
+	state bool automatic = tokens.size() == 2 && tokens[1] == LiteralStringRef("auto");
+	state Reference<ITransaction> tr = db->createTransaction();
+	loop {
+		tr->setOption(FDBTransactionOptions::SPECIAL_KEY_SPACE_ENABLE_WRITES);
+		try {
+			// update cluster description
+			if (new_cluster_description.size()) {
+				tr->set(fdb_cli::clusterDescriptionSpecialKey, new_cluster_description);
+			}
+			// if auto change, read the special key to retrieve the recommended config
+			if (automatic) {
+				// if previous read failed, retry, otherwise, use the same recommened config
+				if (!auto_coordinators_str.size()) {
+					// Hold the reference to the standalone's memory
+					state ThreadFuture<Optional<Value>> auto_coordinatorsF =
+					    tr->get(fdb_cli::coordinatorsAutoSpecialKey);
+					Optional<Value> auto_coordinators = wait(safeThreadFutureToFuture(auto_coordinatorsF));
+					ASSERT(auto_coordinators.present());
+					auto_coordinators_str = auto_coordinators.get().toString();
+				}
+				tr->set(fdb_cli::coordinatorsProcessSpecialKey, auto_coordinators_str);
+			} else if (tokens.size() > 1) {
+				state std::set<NetworkAddress> new_coordinators_addresses;
+				state std::vector<std::string> newAddresslist;
+				state std::vector<StringRef>::iterator t;
+				for (t = tokens.begin() + 1; t != tokens.end(); ++t) {
+					try {
+						auto const& addr = NetworkAddress::parse(t->toString());
+						if (new_coordinators_addresses.count(addr)) {
+							fprintf(stderr, "ERROR: passed redundant coordinators: `%s'\n", addr.toString().c_str());
+							return true;
+						}
+						new_coordinators_addresses.insert(addr);
+						newAddresslist.push_back(addr.toString());
+					} catch (Error& e) {
+						if (e.code() == error_code_connection_string_invalid) {
+							fprintf(
+							    stderr, "ERROR: '%s' is not a valid network endpoint address\n", t->toString().c_str());
+							return true;
+						}
+						throw;
+					}
+				}
+				std::string new_addresses_str = boost::algorithm::join(newAddresslist, ", ");
+				tr->set(fdb_cli::coordinatorsProcessSpecialKey, new_addresses_str);
+			}
+			wait(safeThreadFutureToFuture(tr->commit()));
+			// commit should always fail here
+			// if coordinators are changed, we should get commit_unknown() error
+			ASSERT(false);
+		} catch (Error& e) {
+			state Error err(e);
+			if (e.code() == error_code_special_keys_api_failure) {
+				std::string errorMsgStr = wait(fdb_cli::getSpecialKeysFailureErrorMessage(tr));
+				if (errorMsgStr ==
+				        "Too few fdbserver machines to provide coordination at the current redundancy level" &&
```

Comment:
Same comment as before :), feels like this should be an enum or something instead of string matching.

Reply:
yeah, make sense.
I will make the change.

Reply:
Put these error messages together.
By theory, the best solution for this is to have a special key for it to determine the error type since the CLI version you are using can be different(in most cases, newer than the server version) from the client/server version.
However, these messages are not changed in the last 4 years and we have rare cases to change them.
So I put a comment there if we ever update them, we need to also change the `fdbcli` to catch both the old and the new messages.